src/merkle/index.rs

```diff
@@ -0,0 +1,154 @@
+    pub const fn to_scalar_index(&self) -> u64 {
+        (1 << self.depth as u64) + self.value
+    }
+
+    /// Returns the depth of the current instance.
+    pub const fn depth(&self) -> u8 {
+        self.depth
+    }
+
+    /// Returns the value of the current depth.
+    pub const fn value(&self) -> u64 {
+        self.value
+    }
+
+    /// Returns true if the current value fits the current depth for a binary tree.
+    pub const fn is_valid(&self) -> bool {
+        self.value < (1 << self.depth as u64)
+    }
+
+    /// Returns true if the current instance points to a right sibling node.
+    pub const fn is_value_odd(&self) -> bool {
+        (self.value & 1) == 1
+    }
+
+    /// Returns `true` if the depth is `0`.
+    pub const fn is_root(&self) -> bool {
+        self.depth == 0
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Traverse one level towards the root, decrementing the depth by `1`.
+    pub fn move_up(&mut self) -> &mut Self {
+        self.depth = self.depth.saturating_sub(1);
+        self.value >>= 1;
+        self
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::prelude::*;
+    use rand_utils::prng_array;
+
+    impl NodeIndex {
+        /// Traverse towards the leaves, incrementing the depth by `1`.
+        pub fn traverse(&mut self, right: bool) -> &mut Self {
+            self.depth = self.depth.saturating_add(1);
+            self.value <<= 1;
+            self.value += right as u64;
+            self
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn arbitrary_index_wont_panic_on_traverse(
+            depth in prop::num::u8::ANY,
+            value in prop::num::u64::ANY,
+            count in prop::num::u8::ANY,
+            ref seed in any::<[u8; 32]>()
+        ) {
+            let mut seed = *seed;
+            let mut node = NodeIndex::new(depth, value);
+            for _ in 0..count {
+                seed = prng_array(seed);
+                let args: [u8; 1] = prng_array(seed);
+                let downwards = node.depth() <= 64
+                    && (args[0] & 0b01) == 0b01;
+                let right = (args[0] & 0b10) == 0b10;
+                if downwards {
+                    node.traverse(right);
+                } else {
+                    node.move_up();
+                }
+            }
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn arbitrary_path_traverse_to_valid_index(
+            count in (1_u8..64),
+            ref seed in any::<[u8; 32]>()
+        ) {
+            let mut seed = *seed;
+            let mut node = NodeIndex::root();
+            for _ in 0..count {
+                seed = prng_array(seed);
+                let args: [u8; 1] = prng_array(seed);
+                let right = (args[0] & 1) == 1;
+                node.traverse(right);
+                assert!(node.is_valid());
+            }
+        }
+    }
```

Comment:
Do we need these tests? I'm not quite sure what we are testing here.

Reply:
The only point of these is to check if the provided operations won't panic. Since we removed many of the methods, we are, in practice, currently checking only `move_up` (that it will never panic, it is idempotent, etc)

Reply:
Could we remove references from here to `traverse()` and only test `move_up()` then?

Reply:
Done! It was only a cheap way to move around (up and down) and test more cases.